contract/src/cpmm.rs

```diff
@@ -1,54 +1,143 @@
+use cosmwasm_std::Uint256;
+
 use crate::prelude::*;
 
 #[must_use]
-pub struct UnassignedLpShares(pub LpShare);
-impl UnassignedLpShares {
+pub struct AddLiquidity {
+    pub lp: LpShare,
+    pub returned_to_user: Vec<Token>,
+    pub added_to_pool: Vec<Token>,
+}
+
+impl AddLiquidity {
     pub fn assign_to(
         self,
         storage: &mut dyn Storage,
-        market: &StoredMarket,
+        market: &mut StoredMarket,
         sender: &Addr,
     ) -> Result<()> {
+        assert_eq!(market.outcomes.len(), self.returned_to_user.len());
+        let AddLiquidity {
+            lp,
+            returned_to_user,
+            added_to_pool: _,
+        } = self;
         let mut share_info = ShareInfo::load(storage, market, sender)?
             .unwrap_or_else(|| ShareInfo::new(market.outcomes.len()));
-        share_info.shares += self.0;
+        share_info.shares += lp;
+
+        let mut had_tokens = false;
+        let mut has_tokens = false;
+
+        for (outcome, tokens_mut) in share_info.outcomes.iter_mut().enumerate() {
+            let to_add = returned_to_user[outcome];
+            if tokens_mut.is_zero() {
+                if !to_add.is_zero() {
+                    market.outcomes[outcome].wallets += 1;
+                }
+            } else {
+                had_tokens = true;
+            }
+            *tokens_mut += to_add;
+            if !tokens_mut.is_zero() {
+                has_tokens = true;
+            }
+        }
+
+        if !had_tokens && has_tokens {
+            market.total_wallets += 1;
+        }
```

Comment:
It took me a while to understand this, I wonder if it could be simplified like via an enum:

```
        #[derive(PartialEq)]
        enum Token {
            None,
            Had,
            GotNow,
        }

        let mut token = Token::None;

        for (outcome, tokens_mut) in share_info.outcomes.iter_mut().enumerate() {
            let to_add = returned_to_user[outcome];
            if tokens_mut.is_zero() {
                if !to_add.is_zero() {
                    market.outcomes[outcome].wallets += 1;
                    token = Token::GotNow;
                }
            } else {
                token = Token::Had;
            }
            *tokens_mut += to_add;
        }

        if token == Token::GotNow {
            market.total_wallets += 1;
        }
```

Reply:
I don't think that does the same thing. We need to only increment `market.total_wallets` if _none_ of the outcomes previously had tokens. With this code, if a wallet previously had tokens for outcome 0 and none for 1, we would:

1. Set token to `None` initially
2. Iterate through outcome 0 and set `token` to `Had` (assuming there are no 0 tokens coming in)
3. Iterate through outcome 1 and set `token` to `GotNow`
4. Incorrectly increment `total_wallets`

Reply:
Ah, got it. Thanks!